src/librustc_trans/trans/context.rs

```diff
@@ -977,11 +942,62 @@ fn declare_intrinsic(ccx: &CrateContext, key: &str) -> Option<ValueRef> {
                 ccx.intrinsics().borrow_mut().insert($name, f.clone());
                 return Some(f);
             }
+        );
+        ($name:expr, $cname:ident ($($arg:expr),*) -> $ret:expr) => (
+            if !is_like_pnacl {
+                // The `if key == $name` is already in ifn!
+                ifn!($name, fn($($arg),*) -> $ret);
+            } else if key == $name {
+                let f = declare::declare_cfn(ccx, stringify!($cname),
+                                             Type::func(&[$($arg),*], &$ret),
+                                             ccx.tcx().mk_nil());
+                ccx.intrinsics().borrow_mut().insert($name, f.clone());
+                return Some(f);
+            }
         )
     }
 
     compatible_ifn!("llvm.assume", noop(llvmcompat_assume(i1) -> void), 6);
 
+    // These must be redirected to `libm` here so PNaCl's linker will link in
+    // these functions from `libm`.
+    compatible_ifn!("llvm.copysign.f32", copysignf(t_f32, t_f32) -> t_f32);
```

Comment:
> Also, it won't just work: using __builtin_powi for example and compiling with emcc doesn't error, but still leaves an undefined function llvm_powi_f32, which matches what I found when I look through JSBackend. In order words, LLVM's intrinsics are not handled.

Yes, but that's a bug in Emscripten :) That particular intrinsic (`llvm.powi.*`) was not implemented. Fixed now in https://github.com/kripken/emscripten-fastcomp/commit/fd10012971f5a1143b817002548f0c070cb94498 

> These intrinsics are the only math related intrinsics allow by the PNaCl ABI. All other math-y functions must be linked in via libm (or whatever) during the bitcode link. 

As I understand it, the PNaCl ABI is a definition of the PNaCl **distribution format**, i.e., the thing that ships to browsers. It makes sense to limit that, so that PNaCl executables are as portable as possible. However, there isn't a reason I am aware of to not support all LLVM intrinsics in the **PNaCl developer-side compiler**. If you don't handle `llvm.powi.*` and all other math intrinsics there, then you need hacks in every single compiler targeting PNaCl.